akka-serialization-jackson/src/main/scala/akka/serialization/jackson/JacksonSerializer.scala

```diff
@@ -445,30 +463,32 @@ import akka.util.Helpers.toRootLowerCase
   }
 
   def compress(bytes: Array[Byte]): Array[Byte] = {
-    compressionAlgorithm match {
-      case Compression.Off => bytes
-      case Compression.GZip(largerThan) =>
-        if (bytes.length > largerThan) compressGzip(bytes) else bytes
+    @inline def shouldCompress: Boolean = compressionAlgorithm match {
+      case Compression.Off              => false
+      case Compression.GZip(largerThan) => bytes.length > largerThan
+      case Compression.LZ4(largerThan)  => bytes.length > largerThan
     }
-  }
-
-  private def compressGzip(bytes: Array[Byte]): Array[Byte] = {
+    if (!shouldCompress) return bytes
     val bos = new ByteArrayOutputStream(BufferSize)
-    val zip = new GZIPOutputStream(bos)
+    val zip = compressionAlgorithm match {
+      case Compression.GZip(_) => new GZIPOutputStream(bos)
+      case _                   => new LZ4FrameOutputStream(bos)
```

Comment:
When I tried the performance of this in https://github.com/akka/akka/pull/27131/files#r294805846 it was bad, and that's why we closed that PR. Also after adjusting the block size the results were not great. Generating much garbage. I wonder if that is problem with LZ4FrameOutputStream and we should try to use raw `LZ4Compressor`? Maybe that would work well together with ByteBuffer?

Reply:
I don't think we should use `LZ4Compressor`. I believe `LZ4Compressor` is a bare bone compressor with which the compression result does not contain any metadata. I've confirmed `LZ4Compressor` does not contain magic number by inspecting the outputs of two different buffer. An implementor of this class https://github.com/lz4/lz4-java/blob/1ad67a9845099ca9d4aa10fa18d7d1a631e02fb4/src/java/net/jpountz/lz4/LZ4JNIFastDecompressor.java#L68 uses the c library functions directly. According to the manual http://htmlpreview.github.io/?https://github.com/lz4/lz4/blob/master/doc/lz4_manual.html, the above function does not generate metadata. We need to either add metadata ourselves or use the frame version. The manual states `Each application is free to encode and pass such metadata in whichever way it wants`. We can add our own magic number, compression size encoding etc. I think that would be too much. As for using bytebuffer, I will update the code tomorrow.